### PR TITLE
feat(compass-aggregations): generate aggregations with ai COMPASS-7105

### PR DESCRIPTION
**Pull request by Anemy**
COMPASS-7105

Note: This is using the new backend from https://github.com/10gen/compass-mongodb-com/pull/89/files

This adds the ai input to the aggregation pipeline header. User inputs text, gets an aggregation out. This currently replaces whatever aggregation content they currently have in the aggregation view.

A lot of the actions & reducer code behind fetching the ai response is similar between this pipeline-ai module and the query bar query-ai module. We could move them into a shared package in Compass, I think it would be a bit extra to do that now, and might make things a bit too coupled down the line if we want different experiences in each. Something to keep an eye on.

We currently sample documents every time a generation request is made. We could in future share the preview documents and or cache these documents with the query bar. I opted to keep it simple for now by not implementing that yet.

https://github.com/mongodb-js/compass/assets/1791149/db455f19-4100-470a-afe7-3d2b9f234eda

<details>
  <summary>Screenshots</summary>
  
  Screenshot 2023-08-16 at 1 30 20 PM
  Screenshot 2023-08-16 at 1 31 15 PM
  Screenshot 2023-08-16 at 1 37 17 PM
  Screenshot 2023-08-16 at 2 44 31 PM
</details>

### CODE REVIEW

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx
Comment (Anemy) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx:78]: Will update this to bring in the changes we did to the ai query feedback tests in https://github.com/mongodb-js/compass/pull/4735/commits/d4c12ec432512c9c1302c5fc2e8421ec2a6f8562#diff-d21ac5941571e624a797890f07d97e64bec5852597750e3794ed3716b3c7d471R84

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts
Comment (Anemy) [inline packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts:228]: Linked the wrong Compass ticket here, should be [COMPASS-7100](https://jira.mongodb.org/projects/COMPASS/issues/COMPASS-7100)
Going to likely update this before merging based on when https://github.com/10gen/compass-mongodb-com/pull/89 is merged in.

File: packages/atlas-service/src/main.ts
Comment (Anemy) [inline packages/atlas-service/src/main.ts:589]: I'll add some tests for this function.

File: packages/compass-query-bar/src/utils/query.ts
Comment (mcasimir) [inline packages/compass-query-bar/src/utils/query.ts:54]: `parseQueryAttributesToFormFields` or `parseQueryPropertiesToFormFields`? .. maybe is just me but when i read  parseQueryStrings i can only think that we are parsing a URL

File: packages/atlas-service/src/main.ts
Comment (mcasimir) [inline packages/atlas-service/src/main.ts:650]: seems totally fine to keep that in the backend

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts
Comment (mcasimir) [inline packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts:259]: `LoadGeneratedPipeline`?

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts
Comment (mcasimir) [inline packages/compass-aggregations/src/modules/pipeline-builder/pipeline-ai.ts:270]: more silly naming suggestions: `CancelAIPipelineGenerationAction`?

**Review by mcasimir — COMMENTED** [review]
Epic!! Nothing stood out to me so i only left 1000 naming comments, before we merge it is better if someone else also take a look.

File: packages/atlas-service/src/main.ts
Comment (Anemy) [reply packages/atlas-service/src/main.ts:648]: It's called `AI_MIN_SAMPLE_DOCUMENTS` as this is the length that we slice them to only when the the prompt size is too large with the normal amount of sample documents. So this is the minimum amount we try to send if the normal amount is too much. Happy to name it anything

Comment (mcasimir) [reply packages/atlas-service/src/main.ts:648]: Ah, got it! Something like `AI_FALLBACK_SAMPLE_DOCUMENTS_SIZE`?

File: packages/atlas-service/src/main.ts
Comment (gribnoysup) [reply packages/atlas-service/src/main.ts]: The comment got removed, but I am still not sure why this is a client validation and not a backend one

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx
Comment (gribnoysup) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx:85]: Pipeline AI Feedback?

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/index.tsx
Comment (gribnoysup) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/index.tsx:160]: Is this by design that this is added above the options and not below? Why is it different from query bar?

image

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.tsx
Comment (gribnoysup) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.tsx:34]: This component adds margin that should not be added making it misaligned with all the other content in the toolbar:

image

Layout positioning should be handled by parents, not children, especially for reusable components.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx
Comment (gribnoysup) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:91]: We are already rendering it inside the stages component, why not keep it all in one place and just push it to the right when there are stages in the pipeline?

File: packages/atlas-service/src/main.spec.ts
Comment (gribnoysup) [inline packages/atlas-service/src/main.spec.ts:27]: If they are literally the same to the point that the tests are only switching the endpoint name mock, I wounder if it would make sense to implement them as such and not test the same thing copied over twice

**Review by gribnoysup — APPROVED** [review]
Overall seems to work fine, left a few comments for the implementation.

I am also wondering what is our release strategy for this? As this is not backwards compatible with the currently deployed backend, if we merge this today, this auto creates a beta on Monday early morning and when it releases, latest beta will be broken until we also deploy the backend, right? Do we want to hold off on merging this until beta is kicked off next week or are we okay with the risks?

File: packages/atlas-service/src/main.ts
Comment (alenakhineika) [reply packages/atlas-service/src/main.ts]: We do have this validation on the backend.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx
Comment (alenakhineika) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx:85]: In the compass tracking plan, it is called "AI Response Rated" and it is one kind of event for both query and aggregation that has the `editor_view_type: find | stage | text ` prop to distinguish what we leave feedback for. The plan also uses different values for feedback. It is actually called `rating` with `good | bad | dismissed` values. Should we update this format in both places, here and in the query bar?

```
AI Response Rated

editor_view_type: find | stage | text 
query_shape: <stages & operators used>
rating: good | bad | dismissed
num_tokens: <length of repsonse in number of tokens>
```

**Comment by alenakhineika** [comment]
> I am also wondering what is our release strategy for this? As this is not backwards compatible with the currently deployed backend

@gribnoysup the backend already has the aggregations endpoint. Is this PR uses some different froat from what the endpoint supports?

**Comment by gribnoysup** [comment]
Backend change is merged and autodeployed to staging, but not production as far as I can see, nobody promoted it yet. This is also a breaking change, so if we deploy the backend to production right now for example, the AI feature in the current beta release will stop working because this API responses are not compatible. And vice versa, if we merge this and beta is autopublished on Monday without us releasing the backend in time, AI feature will not work. So that's why I'm asking what's the release strategy for this if we are doing a breaking change without changing the API version in the url and whether or not we should consider waiting until after the kick off just in case (or are okay with timings, risks of doing this right before the kick off)

**Comment by mcasimir** [comment]
> > I am also wondering what is our release strategy for this? As this is not backwards compatible with the currently deployed backend
> 
> @gribnoysup the backend already has the aggregations endpoint. Does this PR use some different format from what the endpoint supports?

Yeah, the change in the service is not released as far as i know, when we merge this one we need to do a new beta, and as soon as the beta is ready we need to promote the service, not the most graceful upgrade, after we are live we should treat breaking changes appropriately.

**Comment by alenakhineika** [comment]
> > > I am also wondering what is our release strategy for this? As this is not backwards compatible with the currently deployed backend
> > 
> > 
> > @gribnoysup the backend already has the aggregations endpoint. Does this PR use some different format from what the endpoint supports?
> 
> Yeah, the change in the service is not released as far as i know, when we merge this one we need to do a new beta, and as soon as the beta is ready we need to promote the service, not the most graceful upgrade, after we are live we should treat breaking changes appropriately.

I think this is where @addaleax suggestion to include the compass version in the request would be useful. Then if we do a breaking change we could still support the previous compass versions.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx
Comment (Anemy) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-ai.spec.tsx:85]: Yes, let's do that in a separate pr.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx
Comment (Anemy) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:91]: I found getting the styles to work nicely without doing it this way ended up being a lot of work. There might be a clean way to do it that's simple but I hadn't found it so I opted for this way.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/index.tsx
Comment (Anemy) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/index.tsx:160]: Nope! Thanks for catching. Pushed up an update:
Screenshot 2023-08-18 at 11 24 39 AM

**Comment by Anemy** [comment]
Going to merge this in once tests go green. As Maurizio mentioned, once this is in main we'll want a new beta and to promote/update the live service accordingly.

**Comment by gribnoysup** [comment]
Adding metadata to the request payload sounds good, but we also have `version` in the endpoint for versioning as specified in the tech design

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx
Comment (gribnoysup) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:91]: Feel free to keep it, sure, just FYI, it's two lines of css as far as I can see, maybe will come up helpful next time you need to style something similarly:

```diff
 const containerStyles = css({
+  flex: 1, // so that the pipeline takes all available space in the toolbar
   display: 'flex',
   gap: spacing[2],
   alignItems: 'center',
 });

+const buttonContainerStyles = css({
+  marginLeft: 'auto', // wrapper for your button so that it's pushed to the right in the layout
+});
```

**Comment by mcasimir** [comment]
> Right, didn’t mean to imply that versioning the API endpoint shouldn’t be the primary way of managing API version differences

To make sure that we have 3 ways to version this API and that we use all of them at once, we should also switch to the atlas convention: `Accept: application/vnd.atlas.<yyyy-mm-dd>+json` (https://www.mongodb.com/docs/atlas/api/versioned-api-overview/)

**Comment by Anemy** [comment]
Merging, I think the versioning and api design changes would be better in a pr separate from this one. Thanks for all the good comments!

**Comment by alenakhineika** [comment]
I will open a PR for the backend to support v1 and v2 endpoints before Monday.

**Pull request merged**
